### ISSUE TITLE
cartesian-product doesn't work for strings

### ISSUE THREAD

**Issue by jonenst** [factor/factor#2392]
```factor
"ab" "de" cartesian-product
! Generic word integer>fixnum does not define a method for the array class.
! Dispatching on object: { 97 100 }

! need to do this...
 "ab" "de" [ >array ] bi@ cartesian-product [ [ >string ] map ] map .
! { { "ad" "ae" } { "bd" "be" } }
```

Something like the following diff would make cartesian-product always return an array of arrays of sequences like the second sequence:

```diff
diff --git a/core/sequences/sequences.factor b/core/sequences/sequences.factor
index f4b7cb2985..cbb6cdb5ab 100644
--- a/core/sequences/sequences.factor
+++ b/core/sequences/sequences.factor
@@ -1088,10 +1088,10 @@ M: repetition sum [ elt>> ] [ length>> ] bi * ; inline
     [ with each ] 2curry each ; inline
 
 : cartesian-map ( ... seq1 seq2 quot: ( ... elt1 elt2 -- ... newelt ) -- ... newseq )
-    [ with map ] 2curry map ; inline
+    [ with { } map-as ] 2curry { } map-as ; inline
 
 : cartesian-product ( seq1 seq2 -- newseq )
-    [ { } 2sequence ] cartesian-map ;
+    dup [ 2sequence ] curry cartesian-map ;
 
 : cartesian-find ( ... seq1 seq2 quot: ( ... elt1 elt2 -- ... ? ) -- ... elt1 elt2 )
     [ f ] 3dip [ with find swap ] 2curry [ nip ] prepose find nip swap ; inline
```
now you directly have
```factor
 "ab" "de" cartesian-product .
! { { "ad" "ae" } { "bd" "be" } } 
```

maybe also add cartesian-product* that use the type of the other sequence ? Add a note in the docs that when the types of the two sequences are not the same, one must be able to hold a superset of the items of the other ?

Maybe add cartesian-product-as for cases where both input sequences types are incompatible:
```factor
"ab" ?{ t f } { } cartesian-product-as .
! { { { 97 t } { 97 f } } { { 98 t } { 98 f } } }
```

so something like this ?
```diff
diff --git a/core/sequences/sequences.factor b/core/sequences/sequences.factor
index f4b7cb2985..4e01875015 100644
--- a/core/sequences/sequences.factor
+++ b/core/sequences/sequences.factor
@@ -1088,10 +1088,16 @@ M: repetition sum [ elt>> ] [ length>> ] bi * ; inline
     [ with each ] 2curry each ; inline
 
 : cartesian-map ( ... seq1 seq2 quot: ( ... elt1 elt2 -- ... newelt ) -- ... newseq )
-    [ with map ] 2curry map ; inline
+    [ with { } map-as ] 2curry { } map-as ; inline
+
+: cartesian-product-as ( seq1 seq2 exemplar -- newseq )
+    [ 2sequence ] curry cartesian-map ;
 
 : cartesian-product ( seq1 seq2 -- newseq )
-    [ { } 2sequence ] cartesian-map ;
+    over cartesian-product-as ;
+
+: cartesian-product* ( seq1 seq2 -- newseq )
+    dup cartesian-product-as ;
 
 : cartesian-find ( ... seq1 seq2 quot: ( ... elt1 elt2 -- ... ? ) -- ... elt1 elt2 )
     [ f ] 3dip [ with find swap ] 2curry [ nip ] prepose find nip swap ; inline
```

**Comment by mrjbq7** [comment]
Some other ideas:

What about defaulting to making arrays, or the exemplar specified so you don’t have the first/second type issue?

: cartesian-map { } cartesian-map-as ;

: cartesian-product { } cartesian-product-as ;

Or perhaps use the type if the types are same, otherwise array?

2dup [ class-of ] same? [ dup ] [ { } ] if cartesian-product-as

**Comment by mrjbq7** [comment]
Fixed in ecd518eddcc2bb9cc8a84ba544ef7c596d303ab1.
